Apps/Inc/hashtable.h

```diff
@@ -0,0 +1,283 @@
+#define HASH_TABLE_STRUCT_NAME CONCAT(HASH_TABLE_NAME, _table_s)
+#define HASH_TABLE_TYPE_NAME CONCAT(HASH_TABLE_NAME, _table_t)
+
+//the struct def for the entries in the hash table
+typedef struct HASH_ENTRY_STRUCT_NAME {
+    HASH_TABLE_KEY_TYPE key;
+    HASH_TABLE_DATA_TYPE data;
+    bool inUse;
+} HASH_ENTRY_TYPE_NAME;
+
+//the struct def for the entire table
+typedef struct HASH_TABLE_STRUCT_NAME {
+    //actual data
+    HASH_ENTRY_TYPE_NAME table[HASH_TABLE_SIZE]; 
+    //other metadata
+    int remaining; //number of remaining elements
+} HASH_TABLE_TYPE_NAME;
+
+/**
+ * @brief Initialize a new hash table
+ * 
+ * If the type of the fifo is new_hashtable, then this function
+ * will be called new_hashtable_init()
+ * 
+ * @return an empty hashtable
+*/
+
+#define INIT CONCAT(HASH_TABLE_NAME, _init)
+static inline HASH_TABLE_TYPE_NAME INIT(){
+    HASH_TABLE_TYPE_NAME table;
+    memset(&table, 0, sizeof(table));
+    table.remaining = HASH_TABLE_SIZE;
+    return table;
+}
+
+/**
+ * @brief Insert an element into the hash table
+ * 
+ * If the type of the fifo is new_hashtable, then this function
+ * will be called new_hashtable_insert()
+ * 
+ * @param table the table to insert into
+ * @param key the key to insert
+ * @param data the data to insert
+ * @return true if the insertion was successful
+*/
+#define INSERT CONCAT(HASH_TABLE_NAME, _insert)
+static inline bool INSERT(HASH_TABLE_TYPE_NAME *table, HASH_TABLE_KEY_TYPE key, HASH_TABLE_DATA_TYPE data){
+    //if the table is full, return false
+    if(table->remaining == 0){
+        return false;
+    }
+    //get the hash of the key using a pearson hash
+    uint8_t hash = 0;
+    uint8_t *keyref = (uint8_t*)&key;
+    for(int i = 0; i < sizeof(key); ++i){
+        hash = lookup[hash ^ keyref[i]];
+    }
+    hash = hash % HASH_TABLE_SIZE;
+
+    //if the hash is already taken, go to next available slot
+    while(table->table[hash].inUse){
+        hash = (hash + 1) % HASH_TABLE_SIZE;
+    }
+    //otherwise, insert the data
+    table->table[hash].key = key;
+    table->table[hash].data = data;
+    table->table[hash].inUse = true;
+    table->remaining--;
+    return true;
+}
+
+/**
+ * @brief Lookup an element in the hash table
+ * 
+ * If the type of the fifo is new_hashtable, then this function
+ * will be called new_hashtable_lookup()
+ * 
+ * @param table the table to lookup in
+ * @param key the key to lookup
+ * @param data pointer to store the result in
+ * @return true if the lookup was successful, false otherwise
+*/
+#define LOOKUP CONCAT(HASH_TABLE_NAME, _lookup)
+static inline bool LOOKUP(HASH_TABLE_TYPE_NAME *table, HASH_TABLE_KEY_TYPE key, HASH_TABLE_DATA_TYPE *data){
+    //get the hash of the key using a pearson hash
+    uint8_t hash = 0;
+    uint8_t *keyref = (uint8_t*)&key;
+    for(int i = 0; i < sizeof(key); ++i){
+        hash = lookup[hash ^ keyref[i]];
+    }
+    hash = hash % HASH_TABLE_SIZE;
+
+    for(int i = 0;i<HASH_TABLE_SIZE;++i){
```

Comment:
This means that a failed lookup is O(n), which is probably not desirable. There's ways to avoid this, such as adding a `deleted` boolean to each bucket; the lookup can halt whenever it reaches a slot that isn't in use and not marked as deleted. The usefulness of this trick however depends on how often we delete elements.